File: lib/core/plugins/sharing/sharing.plugin.dart
import 'package:flutter/services.dart';
import 'package:unisync/core/device.dart';
import 'package:unisync/core/device_connection.dart';
import 'package:unisync/core/plugins/base_plugin.dart';
import 'package:unisync/models/device_message/device_message.model.dart';
import 'package:unisync/utils/logger.dart';
import 'package:url_launcher/url_launcher_string.dart';

class SharingPlugin extends UnisyncPlugin {
  SharingPlugin(Device device)
      : super(device, type: DeviceMessage.Type.SHARING);
  static const _method = (
    OPEN_URL: 'open_url',
    COPY_TEXT: 'copy_text',
  );

  @override
  void onReceive(
      DeviceMessageHeader header, Map<String, dynamic> data, Payload? payload) {
    super.onReceive(header, data, payload);
    if (header.method == _method.OPEN_URL) {
      _handleUrl(data['url']);
    }
    if (header.method == _method.COPY_TEXT) {
      _handleText(data['text']);
    }
  }

  void _handleUrl(String url) {
    debugLog('handling url: $url');
    launchUrlString(url);
  }

  void _handleText(String text) {
    Clipboard.setData(ClipboardData(text: text));
  }
}

File: unisync_android/app/src/main/java/com/anhquan/unisync/core/plugins/sharing/SharingPlugin.kt
package com.anhquan.unisync.core.plugins.sharing

import android.content.Intent
import android.os.Bundle
import android.util.Patterns
import com.anhquan.unisync.core.Device
import com.anhquan.unisync.core.plugins.UnisyncPlugin
import com.anhquan.unisync.models.DeviceMessage

class SharingPlugin(device: Device) :
    UnisyncPlugin(device, DeviceMessage.Type.SHARING) {
    private object Method {
        const val OPEN_URL = "open_url"
        const val COPY_TEXT = "copy_text"
    }

    fun handleExtras(bundle: Bundle) {
        val text = bundle.getString(Intent.EXTRA_TEXT)
        if (text != null) {
            if (isUrl(text)) {
                handleUrl(text)
            } else {
                handleText(text)
            }
        }
    }

    private fun handleUrl(url: String) {
        sendNotification(
            Method.OPEN_URL,
            mapOf(
                "url" to url
            )
        )
    }

    private fun handleText(text: String) {
        sendNotification(
            Method.COPY_TEXT,
            mapOf(
                "text" to text
            )
        )
    }

    private fun isUrl(url: String): Boolean {
        val urlPattern = Patterns.WEB_URL
        val matcher = urlPattern.matcher(url)
        return matcher.matches()
    }
}